### DISCUSSION TITLE
Getting all users with a specific role when using teams

### DISCUSSION THREAD

**Discussion by mohammad-azzam-dev** [spatie/laravel-permission#1999] [Q&A] [answered]
I was using the teams feature in spatie's roles and permissions but I figured out that when using this function:
`Users::role('role1')->get();`

it only returns the users with role1 who are associated with the team id set in the session

my question: how can I get all of the users of different teams?

**Comment by erikn69** [comment, marked as answer]
Make your own relation
```php
// User model
public function roles_all(): BelongsToMany
{
    return $this->morphToMany(
        config('permission.models.role'),
        'model',
        config('permission.table_names.model_has_roles'),
        config('permission.column_names.model_morph_key'),
        PermissionRegistrar::$pivotRole
    );
}
// get all users with a specific role
Users::whereHas('roles_all', fn($query) => $query->whereIn("name", ['role1']))->get();
// or 
Users::whereRelation('roles_all', "name", 'role1')->get();
// Also, you could get all roles, from all teams with this
Users::with('roles_all')->get();
```

**Reply by mohammad-azzam-dev** [reply]
can you please explain what you just did?  or give keywords to search for the functionality of this code

**Reply by erikn69** [reply]
https://laravel.com/docs/8.x/eloquent-relationships#many-to-many-polymorphic-relations
https://laravel.com/docs/8.x/eloquent-relationships#querying-relationship-existence

**Reply by mohammad-azzam-dev** [reply]
thank you

**Reply by mohammad-azzam-dev** [reply]
yes off course I'll try it and get back with you

**Reply by mohammad-azzam-dev** [reply]
it worked, but you should remove the expected return  BelongsToMany and assign it to morphToMany and it works or remove it 

please modify the function so I can set that its the answer and thank you

**Reply by erikn69** [reply]
> remove the expected return BelongsToMany

It's PHP 8 best practice, you're supposed to add the declaration, not remove the return type

https://github.com/spatie/laravel-permission/blob/6a3ed627cee28a552b5176c172ae0abc5eb30925/src/Traits/HasRoles.php#L6
https://github.com/spatie/laravel-permission/blob/6a3ed627cee28a552b5176c172ae0abc5eb30925/src/Traits/HasRoles.php#L43

**Reply by mohammad-azzam-dev** [reply]
ok great and thanks

**Comment by sts-ryan-holton** [comment]
@erikn69 How could I join the "team ID" model to this via my controller? I thought something like this might work but it doesn't:

```php
$user = $user->load([
    'roles_all' => function ($query) {
        $query->join('companies', function ($join) {
            $join->on('roles_all.company_id', '=', 'companies.id');
        });
    }
]);
```
![Screenshot 2023-04-04 at 16 53 36](https://user-images.githubusercontent.com/25177538/229848519-1662cffc-06e3-4684-89b8-78a54dff1dd7.png)

**Reply by erikn69** [reply]
```php
$user = $user->load([
    'roles_all' => function ($query) {
        $query->leftJoin('companies', function ($join) {
            $join->on(config('permission.table_names.roles').'.company_id', '=', 'companies.id');
        });
    }
]);
```

**Reply by erikn69** [reply]
>Any ideas why I don't see my company relationship within the role?

Add a select to the query for get the data,
or if you want the company model make a relationship on Role Model, and nest it

**Comment by sts-ryan-holton** [comment]
@erikn69 Thanks for your reply. I've given this a go, it doesn't give me any errors, this is my code now:

```php
$user = $user->load([
    'roles_all' => function ($query) {
        $query->leftJoin('companies', function ($join) {
            $join->on(config('permission.table_names.roles').'.company_id', '=', 'companies.id');
        });
    }
]);
```

Any ideas why I don't see my `company` relationship within the role?
